### DISCUSSION TITLE
Timeout Error when Calling "onAfterSuiteRun" in "birpc" Module during Test Run

### DISCUSSION THREAD

**Discussion by 201flaviosilva** [vitest-dev/vitest#3457] [Q&A] [answered]
Hey there 👋,

I have been getting this error when running all the tests in my project:

```txt
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Errors ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

Vitest caught 2 unhandled errors during the test run.
This might cause false positive tests. Resolve unhandled errors to make sure your tests are not affected.

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Error ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
Error: [birpc] timeout on calling "onAfterSuiteRun"
 ❯ Timeout._onTimeout node_modules/vitest/dist/vendor-index.3982ff76.js:134:22
 ❯ listOnTimeout node:internal/timers:559:17
 ❯ processTimers node:internal/timers:502:7


⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Error ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
Error: [birpc] timeout on calling "onAfterSuiteRun"
 ❯ Timeout._onTimeout node_modules/vitest/dist/vendor-index.3982ff76.js:134:22
 ❯ listOnTimeout node:internal/timers:559:17
 ❯ processTimers node:internal/timers:502:7

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
```

During the test execution with Vitest, a timeout error occurred when calling the "onAfterSuiteRun" function in the "birpc" module.

I have no idea what that error means 😅

This is my project with the error: https://github.com/201flaviosilva-labs/utilidades/tree/108-use-vite

It's a simple utils lib that I'm being working in my free time. I moved from jest to vitest, and this starts to happen 😅

To reproduce this problem, just clone de repo and change to `108-use-vite` branch and run `npm run test` or `npm run test:dev`.

Any help would be amazing :)
Thanks :)

**Comment by AriPerkkio** [comment]
Might be related to this: https://github.com/vitest-dev/vitest/issues/903

Some test cases are mocking `Math.random`. This is bug.

**Reply by AriPerkkio** [reply, marked as answer]
@sheremet-va should the `birpc` hold original `Math.random` similarly as is does with `setTimeout`, https://github.com/antfu/birpc/pull/7? Patching such change into `node_modules` fixes this issue. Minimal repro to `tests/core/test/mock-internals.test.ts`:

```ts
describe('Math.random', () => {
  beforeEach(() => {
    vi.spyOn(Math, 'random').mockReturnValue(0.1)
  })
  afterEach(() => {
    vi.spyOn(Math, 'random').mockClear()
  })

  test('is mocked', () => {
    expect(Math.random()).toBe(0.1)
  })
})
```

Edit: Actually looks like this works fine even without changing `birpc`:

```diff
+ let spy: SpyInstance
 beforeEach(() => {
-   vi.spyOn(Math, 'random').mockReturnValue(0.1)
+   spy = vi.spyOn(Math, 'random').mockReturnValue(0.1)
 })
 afterEach(() => {
-   vi.spyOn(Math, 'random').mockClear()
+   spy.mockRestore()
 })
```

But by capturing the original `Math.random` in `birpc` the leaking mock would not affect it and make tests stuck. PR open: https://github.com/antfu/birpc/pull/8

**Reply by 201flaviosilva** [reply]
Hey 👋,
Thanks for the help.

I made a repo to create an example: https://bitbucket.org/201flaviosilva-labs/q-vistest-3457/

I changed to adapt to what you did, and it’s still failing 😅 :)

**Reply by AriPerkkio** [reply]
In https://bitbucket.org/201flaviosilva-labs/q-vistest-3457/src/aa3922dfcf3f71ee42e4fb25d19915e9f43fdf10/beep.test.js#lines-4:9 change following:

```diff
describe("choice", () => {
+  let spy;

  beforeEach(() => {
-   vitest.spyOn(global.Math, "random").mockReturnValue(0.1);
+   spy = vitest.spyOn(global.Math, "random").mockReturnValue(0.1);
  });
  afterEach(() => {
-   vitest.spyOn(global.Math, "random").mockRestore();
+   spy.mockRestore();
  });
```

**Reply by 201flaviosilva** [reply]
Sorry, my bad, I had already done this update, but I had not sent it to the remote (and commit).

But the test it's taking a lot of time to run:

```txt
❯ beep.test.js (2)
   ❯ choice (2)
     × should return A
       ⠹ [ afterEach ]
     × should return 10
       ⠹ [ afterEach ]
```

And after that it fails 😅

`spy.spyOn is not a function` :)

**Reply by AriPerkkio** [reply]
The code doesn't match the suggestion above. Change the following:

```diff
afterEach(() => {
-  spy.spyOn(global.Math, "random").mockRestore();
+  spy.mockRestore();
});
```

**Reply by 201flaviosilva** [reply]
🤦 sorry 🤦, sorry for all this, sorry for all the time I made you waste in my distractions 😓

Thanks a lot for your patience and time 😅

**Reply by sheremet-va** [reply]
Did this use case work before? Should we support it?

**Reply by 201flaviosilva** [reply]
The code I had before worked in jest :)

**Reply by sheremet-va** [reply]
We have a bug that needs fixing then.

**Reply by AriPerkkio** [reply]
> Did this use case work before? Should we support it?

The `spyOn().mockClear()` part? No. That doesn't work even in Jest. The only bug Vitest had was that if user does not properly clear `Math.random` mock, the `birpc` will freeze - fixed in #3460.

This does not work in Jest:

```js
test("repro", () => {
  jest.spyOn(global.Math, "random").mockReturnValue(0.5);
  expect(Math.random()).toBe(0.5);


  jest.spyOn(global.Math, "random").mockClear();

  // This will fail as line above doesn't clear the mock.
  expect(Math.random()).not.toBe(0.5);
});
```
